File: minmaxheap/topkfrequent/bigon/TopKFrequentNumbersByBucketing.java
package topkfrequent.bigon;

import java.util.*;

class TopKFrequentNumbersByBucketing {

    public List<Integer> findTopKFrequentNumbers(int[] numbers, int k) {
        // Step 1: Count the frequency of each number
        Map<Integer, Integer> frequencyMap = new HashMap<>();
        for (int number : numbers) {
            frequencyMap.put(number, frequencyMap.getOrDefault(number, 0) + 1);
        }

        // Step 2: Create an array of lists (buckets) to group numbers by frequency
        List<Integer>[] frequencyBuckets = new ArrayList[numbers.length + 1];
        for (int i = 0; i <= numbers.length; i++) {
            frequencyBuckets[i] = new ArrayList<>();
        }

        // Step 3: Place each number into its corresponding frequency bucket
        for (Map.Entry<Integer, Integer> entry : frequencyMap.entrySet()) {
            int number = entry.getKey();
            int frequency = entry.getValue();
            frequencyBuckets[frequency].add(number);
        }

        // Step 4: Collect the top 'K' frequent numbers from the buckets, starting from the highest frequency
        List<Integer> topKFrequentNumbers = new ArrayList<>();
        for (int i = frequencyBuckets.length - 1; i > 0; i--) {  // Start from the highest frequency bucket
            for (int number : frequencyBuckets[i]) {
                topKFrequentNumbers.add(number);
                if (topKFrequentNumbers.size() == k) {  // Stop when we have collected 'K' numbers
                    return topKFrequentNumbers;
                }
            }
        }

        return topKFrequentNumbers;  // This should not be reached if input is valid
    }

    public static void main(String[] args) {
        TopKFrequentNumbersByBucketing finder = new TopKFrequentNumbersByBucketing();

        // Test case 1
        int[] numbers1 = {1, 3, 5, 12, 11, 12, 11};
        int k1 = 2;
        List<Integer> result1 = finder.findTopKFrequentNumbers(numbers1, k1);
        System.out.println("Top K frequent numbers: " + result1);  // Expected output: [12, 11]

        // Test case 2
        int[] numbers2 = {5, 12, 11, 3, 11};
        int k2 = 2;
        List<Integer> result2 = finder.findTopKFrequentNumbers(numbers2, k2);
        System.out.println("Top K frequent numbers: " + result2);  // Expected output: [11, 5] or [11, 12] or [11, 3]
    }
}

File: minmaxheap/topkfrequent/bigon/TopKFrequentNumbersByBucketing.py
from collections import Counter

class TopKFrequentNumbers:
    @staticmethod
    def find_top_k_frequent_numbers(nums, k):
        # Step 1: Count the frequencies of each number
        frequency_map = Counter(nums)

        # Step 2: Initialize an array of empty lists for bucket sorting
        # The size of the bucket is len(nums) + 1 because the max frequency can't exceed the size of nums.
        bucket = [[] for _ in range(len(nums) + 1)]

        # Step 3: Populate the bucket array, where index represents the frequency
        for num, freq in frequency_map.items():
            bucket[freq].append(num)

        # Step 4: Collect the top 'K' frequent numbers from the bucket
        result = []
        for i in range(len(bucket) - 1, 0, -1):  # Iterate from the highest possible frequency to the lowest
            if bucket[i]:
                result.extend(bucket[i])  # Add all numbers with this frequency to the result
            if len(result) >= k:  # Stop once we've collected 'K' numbers
                return result[:k]

        return result

# Example 1
nums = [1, 3, 5, 12, 11, 12, 11]
K = 2
result = TopKFrequentNumbers.find_top_k_frequent_numbers(nums, K)
print("Top K frequent numbers:", result)  # Output: [12, 11]

# Example 2
nums = [5, 12, 11, 3, 11]
K = 2
result = TopKFrequentNumbers.find_top_k_frequent_numbers(nums, K)
print("Top K frequent numbers:", result)  # Output: [11, 5] or [11, 12] or [11, 3]
